### DISCUSSION TITLE
Domain

### DISCUSSION THREAD

**Discussion by mrcsilverfox** [app-anvil/waste_not#60] [Ideas]
# Domain 

## Unità di misura 
`unit`
Quando un prodotto è confezionato singolarmente o può essere trattato come singolo, es. confezione da 12 uova, non è 1 confezione da 12 uova ma item uova da 12 `unit`, ecc.

`liter`
È imoprtante? Quando è fondamentale differenziarla da unit?
Probabilmente solo se l'obiettivo dell'utente è il tracking della propria dispensa.

Se ho 6 confezioni di latte non indicherò 6 litri, perché diventerebbe ingestibile la cosa, dovrei indicare 6 unit di latte.

`kg`
È importante? Quando è fondamentale differenziarla da unit? Simile a `liter`.

Se non prendiamo in considerazione i prodotti "freschi", avere un unit di grana o 500gr di grana poco cambia, se ho 2kg di carne o 1 unit di carne cambia? Perché, se sono tutti insieme?

## Funzionalità

- Cosa vuol dire aggiungere un item?
Aggiungere nella propria dispensa un prodotto con una determinata scadenza, quantità e scegliere dove riporlo. 
È più importante per l'utente avere un item con la quantità aggregata, o tanti item per altrettanti prodotti? In questo caso verrebbe meno la funzionalità unità di misura e sarebbe tutto un unit di quantità 1.
Se aggiungo un nuovo item che ha stesso prodotto, data di scadenza, unità di misura e posizione uguale ad un altro item possiamo sommare le due quantità? Vedremo. 

- Cosa vuol dire aprire un item?
Un item aperto suggerisce all'utente che deve consumarlo al più presto, non si tiene più conto della scadenze effettiva? Ma se lo apro oggi e scade domani, cosa succede? Attualmente l'item sarà segnato come scaduto (e non è presente in quelli aperti).

- Cosa vuol dire consumare completamente un item?
Consumare un item vuol dire rimuoverlo completamente dalla propria dispensa. In assesnza di storico equivale ad un'eliminazione dell'item.

- Cosa vuol dire consumare un item?
Decrementare la quantità rimanente di item, ad esempio se ho un item di 4 unità, posso consumare una sola unità, e dopo avrò in disponsa solamente 3 unità di un determinato item.
Vedremo successivamente se è sempre possibile consumare parzialmente un item. 

L'utente vuole poter indicare come operto un item. 

- Se un item ha come 

| unità di misura / funzionalità  | aprire | consumare | consumare completamente |   |
|---|---|---|---|---|
| litri | ✅ | ❌ | ✅ |
| kg | ✅ | ❌ | ✅ |
| unit  | ✅ | ✅ | ✅ |

Per semplicità un item in `litri` può essere solamente aperto o consumato completamente, non è possibile indicare di aver consumato 100ml di latte.

Un item `kg` potrebbe comportarsi come `litri`, ovvero se viene consumato  equivale ad aprilo, oppure può essere consumato completamente.
Questò perché per consumarlo va aperto ed il resto del item potrebbe andare a male prima della data di scadenza.

Un item `unit` supporta tutte le funzionalità, però se un item ha quantità 3 e viene aperto, cosa significa? Dovrebbe essere creato un nuovo item uguale all'originale, con quantità 1 ed indicato come aperto. 

Deve essere fondamentale avere la flessibilità di spacchettare un item se questo viene aperto se la funzione di merge delle quantità è "attiva".

### Apertura item
Tutti gli item possono essere aperti, indipendentemente dall'unità di misura e dalla quantità. Il senso di aprire un prodotto è ricordare all'utente che c'é un determinato prodotto in dispensa aperto e non del tutto consumato, che andrà a male a breve se non verrà consumato. 
È molto probabile che un utente apra un prodotto per volta, se apre più volte lo stesso item, ogni volta verrà descrementata la unit se di tipo unit. Se apro diversi item con stesso prodotto cosa succede? Creiamo uno singolo item e lo inseriamo negli aperti.

**Comment by gianmarcocalbi** [comment]
> Attualmente l'item sarà segnato come scaduto (e non è presente in quelli aperti)

Questo non va bene. Ho notato che un item che "scade oggi" è segnato come scaduto. In verità le scadenze sono da intendersi "data limite entro cui consumarlo". Su questo dobbiamo fare attenzione perché verrebbe meno tutta la value dell'app, mi spiego meglio, dobbiamo anche essere flessibili sulle scadenze: ad esempio "da consumarsi **preferibilmente** entro" non vuol dire che il giorno dopo è scaduto.

> Per semplicità un item in litri può essere solamente aperto o consumato completamente

Praticamente non esiste più il "consumato parzialmente", se un item ha delle unità puoi modificare le unità, fine. Ad esempio se hai 6 scatole di latte uguali come singolo item con 6 unità, nel momento in apri un item le unità diventano 5 (6-1) e in contemporanea viene creato un nuovo item aperto. Anche perché:

1. l'item aperto può cambiare storage: passare da pantry a fridge come per il caso del latte
2. è giusto mostrare che ci sono 5 pacchi di latte chiusi e uno aperto separatamente e non mostrare un solo item da 6 unità aperto per intero quando è stata aperta una sola unità di latte

> Deve essere fondamentale avere la flessibilità di spacchettare un item se questo viene aperto se la funzione di merge delle quantità è "attiva".

Non ho capito.

> È molto probabile che un utente apra un prodotto per volta, se apre più volte lo stesso item, ogni volta verrà decrementata la unit se di tipo unit.

Esatto.

> Se apro diversi item con stesso prodotto cosa succede? Creiamo uno singolo item e lo inseriamo negli aperti.

No. Creiamo diversi item aperti tenendo conto del giorno in cui sono stati aperti.  
E se uno apre 3 confezioni di latte nello stesso giorno vedrà 3 item diversi aperti nello stesso giorno e va comunque bene.

---

Dobbiamo mettere ben in evidenza gli item "raggruppati", ossia quegli item chiusi con stessa data di scadenza e measure=unit, indicando che si tratta di un gruppo di item e non di un item solo. Per questo cambierei la UI di quando si inserisce un item, non è measure=unit ma mettiamo un'opzione per creare più item identici nello stesso momento. Quindi se scannerizzo un cartone di latte potrò scegliere tra KG/Liter e poi ho l'opzione "Creare 6 item uguali". **Nel dominio cambia l'unità di misura, togliamo "unit", e aggiungiamo un altro contatore separato per le unit,** di default è 1 ma quando "raggruppiamo" gli item avremo un solo item con 6 unit. Ad esempio per il latte avremo un item con:
- measure=L
- quantity=1 (1 litro)
- units=6
